### PR TITLE
May 2022 semantic text search refresh

### PR DESCRIPTION
**Pull request by startakovsky**
@gkogan here is a draft of Semantic Search using a Hugging Face Dataset

### CODE REVIEW

**Comment by gkogan** [comment]
@startakovsky could you please add @jamescalam to this fork so he could take a look too at both PRs?

**Comment by gkogan** [comment]
> @startakovsky could you please add @jamescalam to this fork so he could take a look too at both PRs?

@startakovsky actually you can disregard this, I was able to share a zipped folder with James.

**Comment by gkogan** [comment]
@startakovsky 
* It didn't create an index the first time I ran it; had to uncomment cell 8 and rerun all the following cells.
* Also had to uncomment cell 13
* I couldn't get the upsert to work -- cell 16 just hung at 0% even after 5 minutes.
* Sample query should demonstrate semantic search better. "Eclipse" is just one word and appears in all the result titles, so it looks a lot like keyword search. I couldn't test it out since I got stuck at the upsert step.

**Comment by startakovsky** [comment]
@gkogan There are a couple ways I can speed up the upsert. It's not broken, it's just that there is a lot of text. 
 1) I can sub-sample the documents.
 2) I can limit text to the first few sentences instead of using the entire article.

I will explore both ways. In general, these 700k documents will take some time to encode as vectors, maybe 3 hours.

There is a third way that can assist with perception: I can default to smaller chunks of encoding, which will have the impact of seeing progress sooner, and maybe some impact on total time, hopefully negligible.

Regarding your other comments, I'll work on them.

**Comment by startakovsky** [comment]
@gkogan

TLDR: Feedback addressed. It may take 2-3 hours to complete if you're running it locally on a recent macbook pro. Are you running this on CoLab? It should work there too; in my experience it may take a bit longer for downloads and encodings.

Details:

1) I uncommented the lines you referenced in your comment. 
2) The text is now shortened substantially to speed up 
3) I did not subsample the dataframe so the number of vectors in the index will be 700k. If you want to speed it up just to run the notebook faster than 3 hours, do something like `df = df.sample(100000, random_state=123)`.
4) The notebook will now show a status update in the progress bar for every 32 vectors encoded and then again as vectors get upserted. I have my doubts about this being the fastest (since it's not using multiprocessing), but it is certainly a fast way to see status.
5) I have not yet narrowed down an appropriate sample query, but hopefully you can test it out and see what works. I was kind of thinking "Eclipse car", or something totally different.

image

**Comment by startakovsky** [comment]
@gkogan 

Confirming that when using multiprocessing (my original code) the total time is down to 90 minutes, so there is a tradeoff between providing quick feedback on encoding and total processing time using Sentence Transformers. I invited James to this repository earlier, still waiting on his acceptance else I would @ him too.

image

Detail on computation:
 - note roughly 4 minutes per 20,000 chunk using `model.encode` (screenshot in previous comment) vs 5 minutes per 40,000 using `model.encode_multi_process` (this screenshot)... sentence transformers doesn't come with an easy-to-generate progress bar for `model.encode_multi_process`. It is a feature that I think could be added with some attention.

**Comment by gkogan** [comment]
@startakovsky What if we subsample and add a comment telling user they can toggle the subsample rate if they have more time to spare? For people just kicking the tires during their lunchbreak even 90 mins is a long time to wait.

**Comment by jamescalam** [comment]
I limited the dataset to pull the first 3000 rows from the cc_news dataset, and added a check to make sure the index does not already exist.

If you'd rather sample to keep the data more uniform go for it, I went for the top *n* rows, but imo either approach works. Neither detracts from teaching users about semantic search.

Everything else looks good to me

**Review by jamescalam — COMMENTED** [review]
I limited the dataset to pull the first 3000 rows from the cc_news dataset, and added a check to make sure the index does not already exist.

If you'd rather sample to keep the data more uniform go for it, I went for the top *n* rows, but imo either approach works. Neither detracts from teaching users about semantic search.

Everything else looks good to me
